File: fargo-matchups.py
#!/usr/bin/env python3
"""
Script to calculate race lengths and odds for all player matchups in a FargoRate match.
"""

import argparse
import requests
import sys
from typing import List, Dict, Any, Optional
import matplotlib.pyplot as plt


def get_match(match_id: str) -> Dict[str, Any]:
    """Get match information including teams."""
    url = f"https://lms.fargorate.com/api/matches/{match_id}"
    response = requests.get(url)
    response.raise_for_status()
    return response.json()


def get_team_players(team_id: str) -> List[Dict[str, Any]]:
    """Get list of players for a team."""
    url = f"https://lms.fargorate.com/api/teams/{team_id}/players"
    response = requests.get(url)
    response.raise_for_status()
    return response.json()


def get_races_by_length_and_type(length: str, type_val: int, rating_one: int, rating_two: int) -> Optional[Dict[str, Any]]:
    """Get the closest race for a matchup."""
    url = "https://lms.fargorate.com/api/ratingcalc/racesbylengthandtype"
    params = {
        "length": length,
        "type": type_val,
        "ratingOne": rating_one,
        "ratingTwo": rating_two
    }
    response = requests.get(url, params=params)
    response.raise_for_status()
    races = response.json()

    # Find and return only the race where closest = True
    for race in races:
        if race.get("closest") == True:
            return race

    # If no closest race found, return None
    return None


def get_odds(player_one_rank: int, player_two_rank: int, player_one_race_to: int, player_two_race_to: int) -> Dict[str, Any]:
    """Get odds for a matchup."""
    url = f"https://lms.fargorate.com/api/ratingcalc/odds/{player_one_rank}/{player_two_rank}/{player_one_race_to}/{player_two_race_to}"
    response = requests.get(url)
    response.raise_for_status()
    return response.json()


def get_player_rating(player: Dict[str, Any]) -> int:
    """Extract player rating from player data."""
    # Try different possible field names for rating
    if "rating" in player:
        return int(player["rating"])
    elif "fargoRating" in player:
        return int(player["fargoRating"])
    elif "fargo" in player:
        return int(player["fargo"])
    else:
        # If no rating found, raise an error
        raise ValueError(f"Could not find rating for player: {player}")


def format_odds(odds: Dict[str, Any]) -> str:
    """Format odds data into a compact string."""
    # Try to extract meaningful odds information
    if isinstance(odds, dict):
        # Common odds fields might be: winProbability, odds, etc.
        if "winProbability" in odds:
            prob = odds["winProbability"]
            if isinstance(prob, (int, float)):
                return f"{prob:.1%}"
        if "odds" in odds:
            return str(odds["odds"])
        # If it's a dict, try to get a meaningful value
        values = [str(v) for v in odds.values() if v is not None]
        if values:
            return ", ".join(values[:2])  # Limit to first 2 values
    return str(odds)


def plot_matchup_grid(team1_players: List[Dict[str, Any]], team2_players: List[Dict[str, Any]],
                      matchup_data: List[List[Dict[str, Any]]]):
    """Create a matplotlib visualization of the matchup grid."""
    # Ensure we have exactly 8 players per team (pad if necessary)
    max_players = 8
    team1_display = team1_players[:max_players]
    team2_display = team2_players[:max_players]

    # Pad if needed
    while len(team1_display) < max_players:
        team1_display.append(None)
    while len(team2_display) < max_players:
        team2_display.append(None)

    # Get player names
    team1_names = []
    for p in team1_display:
        if p:
            name = f"{p.get('firstName', '')} {p.get('lastName', '')}".strip()
            team1_names.append(name)
        else:
            team1_names.append("")

    team2_names = []
    for p in team2_display:
        if p:
            name = f"{p.get('firstName', '')} {p.get('lastName', '')}".strip()
            team2_names.append(name)
        else:
            team2_names.append("")

    # Create figure and axis
    fig, ax = plt.subplots(figsize=(16, 12))
    ax.axis('tight')
    ax.axis('off')

    # Prepare table data
    table_data = []
    for i in range(len(team1_names)):
        row = []
        for j in range(len(team2_names)):
            if i < len(matchup_data) and j < len(matchup_data[i]):
                data = matchup_data[i][j]
                if data and "race" in data and "odds" in data:
                    race_str = data["race"]
                    odds_str = format_odds(data["odds"])
                    # Truncate odds if too long
                    if len(odds_str) > 15:
                        odds_str = odds_str[:12] + "..."
                    cell_text = f"{race_str}\n{odds_str}"
                    row.append(cell_text)
                else:
                    row.append("")
            else:
                row.append("")
        table_data.append(row)

    # Create table
    table = ax.table(cellText=table_data,
                     rowLabels=team1_names,
                     colLabels=team2_names,
                     cellLoc='center',
                     loc='center',
                     bbox=[0, 0, 1, 1])

    # Style the table
    table.auto_set_font_size(False)
    table.set_fontsize(9)
    table.scale(1, 2)

    # Get all cells to style them safely
    cells = table.get_celld()

    # Rotate column headers vertically (row 0, columns 1 to len(team2_names))
    for j in range(len(team2_names)):
        pos = (0, j + 1)
        if pos in cells:
            cell = cells[pos]
            cell.set_text_props(rotation=90, ha='center', va='bottom')
            cell.set_facecolor('#E8E8E8')
            cell.set_height(0.1)

    # Style row headers (rows 1 to len(team1_names), column 0)
    for i in range(len(team1_names)):
        pos = (i + 1, 0)
        if pos in cells:
            cell = cells[pos]
            cell.set_facecolor('#E8E8E8')
            cell.set_width(0.15)

    # Style corner cell (0, 0)
    if (0, 0) in cells:
        cells[(0, 0)].set_facecolor('#E8E8E8')

    # Style data cells
    for i in range(len(team1_names)):
        for j in range(len(team2_names)):
            pos = (i + 1, j + 1)
            if pos in cells:
                cell = cells[pos]
                if i < len(matchup_data) and j < len(matchup_data[i]) and matchup_data[i][j]:
                    cell.set_facecolor('#FFFFFF')
                else:
                    cell.set_facecolor('#F5F5F5')
                cell.set_height(0.12)

    # Set title
    plt.title('Matchup Grid: Race and Odds', fontsize=16, fontweight='bold', pad=20)

    # Adjust layout
    plt.tight_layout()

    # Show the plot
    plt.show()


def main():
    parser = argparse.ArgumentParser(description="Calculate race lengths and odds for all player matchups in a match")
    parser.add_argument("--match-id", required=True, help="The match ID to analyze")
    args = parser.parse_args()

    try:
        # Get match information
        print(f"Fetching match {args.match_id}...")
        match = get_match(args.match_id)

        team1_id = match["teamOneId"]
        team2_id = match["teamTwoId"]

        print(f"Team 1 ID: {team1_id}")
        print(f"Team 2 ID: {team2_id}")

        # Get players for each team
        print(f"\nFetching players for team 1...")
        team1_players = get_team_players(team1_id)
        print(f"Found {len(team1_players)} players")

        print(f"\nFetching players for team 2...")
        team2_players = get_team_players(team2_id)
        print(f"Found {len(team2_players)} players")

        print(f"\nAnalyzing {len(team1_players)} x {len(team2_players)} potential matchups...")

        # Collect matchup data in a 2D array
        matchup_data = []
        for p1 in team1_players:
            row = []
            for p2 in team2_players:
                try:
                    rating1 = get_player_rating(p1)
                    rating2 = get_player_rating(p2)

                    # Determine race length and type based on player ratings
                    # If both players are under 400: length = 3, type = 2 (hot)
                    # Otherwise: length = 4, type = 1 (medium)
                    if rating1 < 400 and rating2 < 400:
                        length = "3"
                        type_val = 2
                    else:
                        length = "4"
                        type_val = 1

                    try:
                        race = get_races_by_length_and_type(length, type_val, rating1, rating2)

                        if race:
                            high_player_race_to = race.get("highPlayerRaceTo")
                            low_player_race_to = race.get("lowPlayerRaceTo")

                            # Determine which player is higher rated
                            if rating1 >= rating2:
                                p1_race_to = high_player_race_to
                                p2_race_to = low_player_race_to
                            else:
                                p1_race_to = low_player_race_to
                                p2_race_to = high_player_race_to

                            # Get odds for this race
                            try:
                                odds = get_odds(rating1, rating2, p1_race_to, p2_race_to)
                                row.append({
                                    "race": f"{p1_race_to}-{p2_race_to}",
                                    "odds": odds
                                })
                            except Exception as e:
                                row.append({
                                    "race": f"{p1_race_to}-{p2_race_to}",
                                    "odds": f"Error: {e}"
                                })
                        else:
                            row.append(None)

                    except Exception as e:
                        row.append({"race": "Error", "odds": str(e)})

                except Exception as e:
                    row.append({"race": "Error", "odds": str(e)})

            matchup_data.append(row)

        # Create and display the grid plot
        plot_matchup_grid(team1_players, team2_players, matchup_data)

    except requests.exceptions.RequestException as e:
        print(f"Error making API request: {e}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    main()

File: fargo-matchups.js
#!/usr/bin/env node
/**
 * Script to calculate race lengths and odds for all player matchups in a FargoRate match.
 */

import axios from 'axios';
import { writeFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

/**
 * Get match information including teams.
 */
async function getMatch(matchId) {
    const url = `https://lms.fargorate.com/api/matches/${matchId}`;
    const response = await axios.get(url);
    return response.data;
}

/**
 * Get list of players for a team.
 */
async function getTeamPlayers(teamId) {
    const url = `https://lms.fargorate.com/api/teams/${teamId}/players`;
    const response = await axios.get(url);
    return response.data;
}

/**
 * Get the closest race for a matchup.
 */
async function getRacesByLengthAndType(length, typeVal, ratingOne, ratingTwo) {
    const url = "https://lms.fargorate.com/api/ratingcalc/racesbylengthandtype";
    const params = {
        length: length,
        type: typeVal,
        ratingOne: ratingOne,
        ratingTwo: ratingTwo
    };
    const response = await axios.get(url, { params });
    const races = response.data;

    // Find and return only the race where closest = true
    for (const race of races) {
        if (race.closest === true) {
            return race;
        }
    }

    // If no closest race found, return null
    return null;
}

/**
 * Get odds for a matchup.
 */
async function getOdds(playerOneRank, playerTwoRank, playerOneRaceTo, playerTwoRaceTo) {
    const url = `https://lms.fargorate.com/api/ratingcalc/odds/${playerOneRank}/${playerTwoRank}/${playerOneRaceTo}/${playerTwoRaceTo}`;
    const response = await axios.get(url);
    return response.data;
}

/**
 * Extract player rating from player data.
 */
function getPlayerRating(player) {
    if (player.rating !== undefined) {
        return parseInt(player.rating);
    } else if (player.fargoRating !== undefined) {
        return parseInt(player.fargoRating);
    } else if (player.fargo !== undefined) {
        return parseInt(player.fargo);
    } else {
        throw new Error(`Could not find rating for player: ${JSON.stringify(player)}`);
    }
}

/**
 * Format odds data into a compact string.
 */
function formatOdds(odds) {
    if (typeof odds === 'object' && odds !== null) {
        if (odds.winProbability !== undefined) {
            const prob = odds.winProbability;
            if (typeof prob === 'number') {
                return `${(prob * 100).toFixed(1)}%`;
            }
        }
        if (odds.odds !== undefined) {
            return String(odds.odds);
        }
        // If it's an object, try to get meaningful values
        const values = Object.values(odds).filter(v => v !== null && v !== undefined).map(String);
        if (values.length > 0) {
            return values.slice(0, 2).join(', ');
        }
    }
    return String(odds);
}

/**
 * Create an HTML visualization of the matchup grid.
 */
function createMatchupGridHTML(team1Players, team2Players, matchupData) {
    const maxPlayers = 8;
    const team1Display = team1Players.slice(0, maxPlayers);
    const team2Display = team2Players.slice(0, maxPlayers);

    // Pad if needed
    while (team1Display.length < maxPlayers) {
        team1Display.push(null);
    }
    while (team2Display.length < maxPlayers) {
        team2Display.push(null);
    }

    // Get player names
    const team1Names = team1Display.map(p => {
        if (p) {
            const name = `${p.firstName || ''} ${p.lastName || ''}`.trim();
            return name;
        }
        return '';
    });

    const team2Names = team2Display.map(p => {
        if (p) {
            const name = `${p.firstName || ''} ${p.lastName || ''}`.trim();
            return name;
        }
        return '';
    });

    // Build table rows
    let tableRows = '';
    for (let i = 0; i < team1Names.length; i++) {
        tableRows += '<tr>';
        // Row header
        tableRows += `<th class="row-header">${team1Names[i] || ''}</th>`;

        // Data cells
        for (let j = 0; j < team2Names.length; j++) {
            if (i < matchupData.length && j < matchupData[i].length && matchupData[i][j]) {
                const data = matchupData[i][j];
                const race = data.race || '';
                const odds = formatOdds(data.odds || '');
                const oddsDisplay = odds.length > 15 ? odds.substring(0, 12) + '...' : odds;
                tableRows += `<td class="data-cell"><div class="race">${race}</div><div class="odds">${oddsDisplay}</div></td>`;
            } else {
                tableRows += '<td class="data-cell empty"></td>';
            }
        }
        tableRows += '</tr>';
    }

    // Build column headers
    let colHeaders = '<tr><th class="corner"></th>';
    for (const name of team2Names) {
        colHeaders += `<th class="col-header"><div class="rotated">${name || ''}</div></th>`;
    }
    colHeaders += '</tr>';

    const html = `<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Matchup Grid: Race and Odds</title>
    <style>
        * {
            margin: 0;
            padding: 0;
            box-sizing: border-box;
        }

        body {
            font-family: Arial, sans-serif;
            padding: 20px;
            background-color: #f5f5f5;
        }

        h1 {
            text-align: center;
            margin-bottom: 20px;
            color: #333;
        }

        .container {
            background-color: white;
            padding: 20px;
            border-radius: 8px;
            box-shadow: 0 2px 4px rgba(0,0,0,0.1);
            overflow-x: auto;
        }

        table {
            border-collapse: collapse;
            width: 100%;
            margin: 0 auto;
        }

        th, td {
            border: 1px solid #ddd;
            padding: 8px;
            text-align: center;
            min-width: 120px;
        }

        .corner {
            background-color: #E8E8E8;
            width: 150px;
        }

        .row-header {
            background-color: #E8E8E8;
            font-weight: bold;
            text-align: right;
            padding-right: 15px;
            width: 150px;
        }

        .col-header {
            background-color: #E8E8E8;
            height: 150px;
            vertical-align: bottom;
            padding: 5px;
        }

        .rotated {
            transform: rotate(-90deg);
            transform-origin: center;
            white-space: nowrap;
            display: inline-block;
            width: 140px;
            text-align: left;
        }

        .data-cell {
            background-color: #FFFFFF;
            height: 80px;
            vertical-align: middle;
        }

        .data-cell.empty {
            background-color: #F5F5F5;
        }

        .race {
            font-weight: bold;
            font-size: 14px;
            margin-bottom: 5px;
        }

        .odds {
            font-size: 12px;
            color: #666;
        }
    </style>
</head>
<body>
    <h1>Matchup Grid: Race and Odds</h1>
    <div class="container">
        <table>
            ${colHeaders}
            ${tableRows}
        </table>
    </div>
</body>
</html>`;

    return html;
}

/**
 * Main function
 */
async function main() {
    const args = process.argv.slice(2);
    const matchIdIndex = args.indexOf('--match-id');

    if (matchIdIndex === -1 || matchIdIndex === args.length - 1) {
        console.error('Error: --match-id argument is required');
        process.exit(1);
    }

    const matchId = args[matchIdIndex + 1];

    try {
        // Get match information
        console.log(`Fetching match ${matchId}...`);
        const match = await getMatch(matchId);

        const team1Id = match.teamOneId;
        const team2Id = match.teamTwoId;

        console.log(`Team 1 ID: ${team1Id}`);
        console.log(`Team 2 ID: ${team2Id}`);

        // Get players for each team
        console.log('\nFetching players for team 1...');
        const team1Players = await getTeamPlayers(team1Id);
        console.log(`Found ${team1Players.length} players`);

        console.log('\nFetching players for team 2...');
        const team2Players = await getTeamPlayers(team2Id);
        console.log(`Found ${team2Players.length} players`);

        console.log(`\nAnalyzing ${team1Players.length} x ${team2Players.length} potential matchups...`);

        // Collect matchup data in a 2D array
        const matchupData = [];
        for (const p1 of team1Players) {
            const row = [];
            for (const p2 of team2Players) {
                try {
                    const rating1 = getPlayerRating(p1);
                    const rating2 = getPlayerRating(p2);

                    // Determine race length and type based on player ratings
                    // If both players are under 400: length = 3, type = 2 (hot)
                    // Otherwise: length = 4, type = 1 (medium)
                    let length, typeVal;
                    if (rating1 < 400 && rating2 < 400) {
                        length = "3";
                        typeVal = 2;
                    } else {
                        length = "4";
                        typeVal = 1;
                    }

                    try {
                        const race = await getRacesByLengthAndType(length, typeVal, rating1, rating2);

                        if (race) {
                            const highPlayerRaceTo = race.highPlayerRaceTo;
                            const lowPlayerRaceTo = race.lowPlayerRaceTo;

                            // Determine which player is higher rated
                            let p1RaceTo, p2RaceTo;
                            if (rating1 >= rating2) {
                                p1RaceTo = highPlayerRaceTo;
                                p2RaceTo = lowPlayerRaceTo;
                            } else {
                                p1RaceTo = lowPlayerRaceTo;
                                p2RaceTo = highPlayerRaceTo;
                            }

                            // Get odds for this race
                            try {
                                const odds = await getOdds(rating1, rating2, p1RaceTo, p2RaceTo);
                                row.push({
                                    race: `${p1RaceTo}-${p2RaceTo}`,
                                    odds: odds
                                });
                            } catch (error) {
                                row.push({
                                    race: `${p1RaceTo}-${p2RaceTo}`,
                                    odds: `Error: ${error.message}`
                                });
                            }
                        } else {
                            row.push(null);
                        }
                    } catch (error) {
                        row.push({ race: "Error", odds: error.message });
                    }
                } catch (error) {
                    row.push({ race: "Error", odds: error.message });
                }
            }
            matchupData.push(row);
        }

        // Create HTML visualization
        const html = createMatchupGridHTML(team1Players, team2Players, matchupData);
        const outputPath = join(__dirname, 'matchup-grid.html');
        writeFileSync(outputPath, html, 'utf8');

        console.log(`\nVisualization saved to ${outputPath}`);

        // Try to open in browser
        try {
            const { default: open } = await import('open');
            console.log('Opening in browser...');
            await open(outputPath);
        } catch (error) {
            console.log(`Please open ${outputPath} in your browser to view the visualization.`);
        }

    } catch (error) {
        if (error.response) {
            console.error(`Error making API request: ${error.message}`, error.response.data);
        } else {
            console.error(`Error: ${error.message}`);
            console.error(error.stack);
        }
        process.exit(1);
    }
}

// Run main function
main();

